# Practice 3 TAB: Pacman with AlphaBeta + Neural Networks

### *Autores:* ***Sergio Azorín Julián y José Francisco Hurtado***

## **1. Introducción**

Esta práctica tiene como propósito fundamental diseñar e implementar un agente capaz de jugar a Pacman de una manera inteligente, integrando estrategias de anticipación al adverario (AlphaBeta) con modelos de aprendizaje automático (Red Neuronal).

Para la realización de la práctica hemos llevado a cabo un proceso evolutivo y acumulativo, en el que partimos de agentes que empleaban puramente la intuición de la red neuronal hacia sistemas híbridos que combinan el experiencia de las partidas previas con planificaciones estategicas que analizan la partida actual unos pasos por delante.

Iniciamos nuestro trabajo a partir de un agente neuronal puro (en nuestro caso NeuralAgentDummy), un agente entrenado a partir de un conjunto de partidas jugadas por nosotros. Esta forma de actuar se basaba únicamente en evaluar el tablero actual y seleccionar la acción con mayor probabilidad con respecto a su entrenamiento. Sin embargo, al no presentar ningún tipo de estrategia, este agente acaba siendo incapaz de prever trampas a corto, medio y largo plazo.

Con el fin de mitigar los fallos de supervivencia de nuestro modelo neuronal, se ha implementado un segundo agente en el que se complementa la salida de la red neuronal con un conjunto de heurísticas basadas, por ejemplo, en la distacia a la comida y a los fantasmas. A pesar de otorgar una mejora sobre el anterior agente, sigue manteniendo el mismo problema, no es capaz de anticiparse a los movimientos de los enemigos, teniendo un comportamiento puramente reactivo (greedy, horizonte de búsqueda de 0).

Para soluciónar esta cuestión se ha optado por emplear Minimax, un método de busqueda adversarial, el cual, a partir de un árbol que represente el conjunto de movimientos legales, podrémos encontrar la acción que más beneficia a Pacman suponiendo que los adversarios van a jugar óptimamente. Debemos de tener en cuenta que este tipo de busquedas presentan un coste computacional exponenecial a medida que aumentamos la profundidad de busqueda. Por ello, utilizaremos AlphaBeta como técnica de optimización, la cual reducirá el conjunto de nodos evaluados sin afectar al resultado final.

Con este nuevo tipo de técnicas somos capaces de superar el enfoque reactivo, otorgándole a nuestro agente inteligente la posibilidad de crear estrategias. Este agente definitivo nace con el proposito de fusionar la experiencia acumulada por la red neuronal y la anticipación a las jugadas del rival.

Para consolidar esta fusión de forma robusta, el agente implementa dos innovaciones clave:

- **Función de Evaluación Híbrida**: Que unifica en un solo valor las estimaciones de la red neuronal y las reglas heurísticas.

- **Estrategia de pesos dinámicos**: Una mejora que varía la importacia de cada componente según el progreso de la partida.

A través de este documento, se expondrá en mayor detalle la implementación de este agente definitivo (AlphaBetaNeuralAgent) y los resultados que hemos obtenido al otrogarle el control de la partida.

## Parte 1. Nuevas Heurísticas

### 1. Introducción
#### ¿Por qué hacen falta las heurísticas?
Partimos de un problema troncal en cualquier juego, si quieres ganar partidas de manera consistente (que no sea suerte / azar) necesitas anticiparte a lo que tu rival va a hacer. Eso implica imaginar lo que tu adversario va a hacer en cada movimiento, y en muchas ocasiones hasta varios en movimientos. El inconveniente que esto acarrea es el coste combinatorial, existen una cantidad exponencial de posibles jugadas que se pueden dar partiendo de un moviento y anticipando 6 o 7 jugadas. En otras palabras, es muy fácil omitir un caso que acabe siendo catastrófico y acabar perdiendo o tener que calcularlo todo es extremadamente lento e inviable computacionalmente hablando. En resumen, tratar de aplicar la fuerza bruta no es suficiente para poder ganar de manera consistente.   

#### La solución: Usar heurísticas
La utilización de las heurísticas, estas son como "intuiciones" matemáticas que tiene el agente para poder calcular / evaluar si la situación en la que se encuentra le favorece a él o al rival ¿cómo? pues teniendo una lista de directrices (buenas jugadas / costumbres) que le indican si le benefician a él o no.

#### Aplicado al Pacman
En el caso del Pacman nos vamos a basar en factores relacionados con la obtención de la comida, evasión de los fantasmas y estrategias para evitar exponerse a situaciones vulnerables como quedar acorralado en un callejón por ambos fantasmas. Dependiendo de estos factores se irán sumando o restando puntos a una variable que será el score de la heurística que será el que indique como de favorable o desfavorable es la posición en cuestión para Pacman. Cuando mayor sea el score mejor será la situación para él y viceversa, cuanto peor score peor situación.

### 2. Extracción de los datos de la partida
Antes de empezar a aplicar las heurísticas necesitamos saber en que escenario se encuentra Pacman, por eso obtenemos todos los datos relevantes del estado de la partida. Es lo primero que se hace al llamar a nuestra función `traditional_evaluation`: 

#### Código

```Python
    score = state.getScore()
    pacman_pos = state.getPacmanPosition()
    food = state.getFood().asList()
    ghost_states = state.getGhostStates()
    capsules = state.getCapsules()
    legal_actions = state.getLegalActions()
```

#### ¿Qué es cada cosa?
- Score: Es la puntuación base que nos da el propio juego. La usamos como punto de partida. Sobre esta puntuación base, nosotros sumaremos o restaremos nuestros propios puntos de heurística para "guiar" a Pacman.

- pacman_pos: Es la coordenada exacta de nuestro Pacman (ej: x=5, y=3). Es nuestro punto de referencia para medir a qué distancia están las cosas buenas (comida) y las cosas malas (fantasmas).

- food: Aquí sacamos dónde están las bolitas de comida. Usamos .asList() para convertir el mapa del juego en una simple lista de coordenadas. Así es mucho más fácil y rápido calcular distancias luego.

- ghost_states: Este dato es clave, no solo obtenemos la posición de los fantasmas, sino su estado. En Pacman, un fantasma puede ser mortal (hay que huir) o puede estar asustado (hay que comérselo). Esta variable nos dirá si el temporizador de miedo (scaredTimer) está activo o no.

- capsules: Las coordenadas de las cápsulas.

- legal_actions: Nos dice qué movimientos puede hacer Pacman en este preciso lugar (Norte, Sur, Este, etc.)

#### ¿Cómo medimos las distancias entre objetos?
Para calcular cómo de lejos está Pacman de la comida o de los fantasmas, usamos la Distancia de Manhattan. Esta mide los pasos reales que hay que dar moviéndose en forma de cruz (arriba, abajo, izquierda, derecha).

## Parte 2. Entrenamiento de la Red Neuronal

## Parte 3. AlphaBetaNeuralAgent

En esta sección se detalla la implementación del AlphaBetaNeuralAgent, el cual logra unificar el razonamiento adversarial a múltiples pasos (mediante la Poda Alfa-Beta) con la intuición predictiva de un modelo de aprendizaje profundo.

In [ ]:
class AlphaBetaNeuralAgent(AlphaBetaAgent):
    
    def __init__(self,
                 evalFn='scoreEvaluationFunction',
                 depth='3',
                 start_heuristicsWeight=0.3,  # Peso de AlphaBeta al INICIO
                 start_nnWeight=0.7,          # Peso de la Red Neuronal al INICIO
                 end_heuristicsWeight=0.7,    # Peso de AlphaBeta al FINAL
                 end_nnWeight=0.3):           # Peso de la Red Neuronal al FINAL):
        super().__init__(evalFn, depth)

        self.evaluationFunction = self.evaluation_combined
        
        # Guardamos los límites de la transición
        self.start_w_heuristic = start_heuristicsWeight
        self.start_w_neural = start_nnWeight
        self.end_w_heuristic = end_heuristicsWeight
        self.end_w_neural = end_nnWeight
        
        # Pesos actuales que usará evaluation_combined
        self.w_heuristic = self.start_w_heuristic
        self.w_neural = self.start_w_neural
        
        self.neural_agent_dummy = NeuralAgentDummy()
        
        # Guardaremos la cantidad inicial de comida para calcular el progreso
        self.initial_food = None
    
    def evaluation_combined(self, state):
        # 1) Traditional score (with the new heuristics from Task 1)
        trad_score = traditional_evaluation(state)
        #print(f"Heuristic score: {trad_score}")

        # 2) Neural network score
        neural_score = self.neural_agent_dummy.neural_evaluation(state)
        #print(f"Neural network score: {neural_score}")

        # 3) Weighted combination
        return self.w_heuristic * trad_score + self.w_neural * neural_score


    def getAction(self, gameState: GameState):
        """
        Returns the minimax action using self.depth and self.evaluationFunction
        """
        "*** YOUR CODE HERE ***"
        # 1. Registrar la comida inicial en el primer turno
        if self.initial_food is None:
            self.initial_food = gameState.getNumFood()
            # Si el mapa empieza sin comida (raro, pero previene división por cero)
            if self.initial_food == 0:
                self.initial_food = 1 

        # 2. Calcular el progreso del juego (de 0.0 al inicio, a 1.0 al final)
        current_food = gameState.getNumFood()
        progress = 1.0 - (current_food / self.initial_food)

        # 3. Interpolación lineal de los pesos
        self.w_heuristic = self.start_w_heuristic * (1 - progress) + self.end_w_heuristic * progress
        self.w_neural = self.start_w_neural * (1 - progress) + self.end_w_neural * progress


        def alphabeta(agentIndex, depth, gameState, alpha, beta):

            if (gameState.isWin() or
                gameState.isLose() or
                depth == self.depth):
                return self.evaluationFunction(gameState)

            #Max (Pacman)
            if agentIndex == 0:
                return maxValue(agentIndex,depth,gameState,alpha,beta)

            #Min (Fantasmas)
            return minValue(agentIndex,depth,gameState,alpha,beta)

        def maxValue(agentIndex,depth,gameState,alpha,beta):

            v = float('-inf')
            legalActions = gameState.getLegalActions(agentIndex)

            if not legalActions:
                return self.evaluationFunction(gameState)

            for action in legalActions:
                successor = gameState.generateSuccessor(agentIndex,action)
                value = alphabeta(1,depth,successor,alpha,beta)
                v = max(v,value)

                #Hacemos poda
                if v >= beta:
                    return v

                alpha = max(alpha, v)

            return v


        def minValue(agentIndex,depth,gameState,alpha,beta):
            v = float('inf')


            legalActions = gameState.getLegalActions(agentIndex)

            if not legalActions:
                return self.evaluationFunction(gameState)

            nextAgent = agentIndex+1
            nextDepth = depth

            if nextAgent == gameState.getNumAgents():
                nextAgent = 0
                nextDepth = depth + 1

            for action in legalActions:

                successor = gameState.generateSuccessor(agentIndex,action)

                value= alphabeta(nextAgent,nextDepth,successor,alpha,beta)

                v = min(v,value)

                #Hacemos poda
                if v <= alpha:
                    return v

                beta = min(beta, v)

            return v

        alpha = float('-inf')
        beta = float('inf')

        bestAction = None
        bestScore = float('-inf')

        for action in gameState.getLegalActions(0):
            successor = gameState.generateSuccessor(0,action)

            score = alphabeta(1,0,successor,alpha,beta)

            if score > bestScore:
                bestScore = score
                bestAction = action

            alpha = max(alpha,bestScore)

        return bestAction

**Función de Evaluación Combinada**

Para evaluar la utilidad de los estados futuros, el agente no se fía de una única fuente. En su lugar, utiliza el método evaluation_combined, que realiza una suma ponderada de dos componentes principales:

- Puntuación Tradicional (trad_score): Evalúa el estado basándose en el análisis explícito de características del entorno (distancia a la comida, proximidad de los fantasmas y las nuevas heurísticas introducidas en la Parte 1).

- Puntuación Neuronal (neural_score): Aporta el conocimiento inferido por la red neuronal tras su entrenamiento con partidas victoriosas.

Ambos valores se multiplican por sus respectivos pesos (w_heuristic y w_neural) para dictaminar el valor final del nodo.

**Estrategia de Pesos Dinámicos**

Como mejora avanzada sobre los pesos estáticos y con el objetivo de maximizar el rendimiento , se ha diseñado una estrategia de pesos dinámicos dependientes del progreso de la partida. La implementación funciona de la siguiente manera: 

- Métrica de progreso: En el primer movimiento se registra la cantidad total de comida en el nivel. En cada turno, se calcula un factor de progreso normalizado de $0.0$ a $1.0$ basado en la comida consumida (progress = 1.0 - (current_food / initial_food)).

- Transición de prioridades: Al comienzo del nivel (progreso cercano a $0.0$), el agente da mayor peso a la intuición general de la red neuronal (start_nnWeight = 0.7) para explorar y trazar rutas globales de forma eficiente.

- Fase final (Endgame): A medida que la comida escasea (progreso cercano a $1.0$), se realiza una interpolación lineal de los parámetros para transferir el control progresivamente a las heurísticas analíticas (end_heuristicsWeight = 0.7).

Durante la ejecución de diversas pruebas con diferentes pesos hemos decidido otorgarle mayor peso a AlphaBeta + Heuristicas al final de la partida, con el objetivo de que asegure la victoria. Esto se debe a que hemos notado que si le damos el control a la red neuronal, Pacman actua de manera más agresiva consumiendo gran cantidad de comida sin preocuparse en exceso por su supervivencia, lo cual puede ser algo positivo al inicio, sin embargo, a medida que nos vamos acercando al endgame, comenzamos a encontrarnos restos de comida esparcidos por todo el mapa. Para esta situación de endgame, hemos decidido darle el control a AlphaBeta, ya que juega más tranquilo y asegura la supervivencia de Pacman mucho mejor de lo que lo hace la Red Neuronal. En general, AlphaBeta es más lento completando el objetivo, pero las decisiones que toma son más seguras.

## Parte 4. Resultados

In [9]:
from texttable import Texttable
import statistics



old_layout_Greedy=[-436.0, -461.0, -477.0, -410.0, -85.0, -465.0, -478.0, -316.0, -276.0, -463.0]
new_layout_Greedy=[-459.0, -467.0, -455.0, -447.0, -192.0, -264.0, -463.0, -213.0, -402.0, -121.0]

old_layout_GreedyH=[-376.0, -368.0, 1650.0, 777.0, 533.0, 2031.0, 784.0, 167.0, 117.0, 1053.0]
new_layout_GreedyH=[-376.0, -368.0, 975.0, 412.0, 1510.0, 827.0, 440.0, -349.0, 606.0, 1644.0]

old_layout_AlphaBeta=[2053.0, 1872.0, 1899.0, 662.0, 1909.0, 1484.0, 2068.0, -351.0, 403.0, 945.0]
new_layout_AlphaBeta=[445.0, 1829.0, 1825.0, 1895.0, -367.0, 671.0, 1731.0, 1092.0, 874.0, 1727.0]


table = Texttable()
table.add_rows([["Configuration", "Classical layout", "New layout"],
               ["Greedy neural agent",f"score: {statistics.mean(old_layout_Greedy)} \nwin_rate: 0/10",f"score: {statistics.mean(new_layout_Greedy)} \nwin_rate: 0/10"],
               ["Greedy neural agent + heuristics",f"score: {statistics.mean(old_layout_GreedyH)} \nwin_rate: 2/10",f"score: {statistics.mean(new_layout_GreedyH)} \nwin_rate: 2/10"],
               ["AlphaBeta + NN + heuristics",f"score: {statistics.mean(old_layout_AlphaBeta)} \nwin_rate: 6/10",f"score: {statistics.mean(new_layout_AlphaBeta)} \nwin_rate: 5/10"]])
print(table.draw())

+----------------------------------+------------------+----------------+
|          Configuration           | Classical layout |   New layout   |
+==================================+==================+================+
| Greedy neural agent              | score: -386.7    | score: -348.3  |
|                                  | win_rate: 0/10   | win_rate: 0/10 |
+----------------------------------+------------------+----------------+
| Greedy neural agent + heuristics | score: 636.8     | score: 532.1   |
|                                  | win_rate: 2/10   | win_rate: 2/10 |
+----------------------------------+------------------+----------------+
| AlphaBeta + NN + heuristics      | score: 1294.4    | score: 1172.2  |
|                                  | win_rate: 6/10   | win_rate: 5/10 |
+----------------------------------+------------------+----------------+


A partir de la tabla comparativa generada tras la ejecución de al menos 10 partidas por configuración, se pueden extraer conclusiones muy claras sobre el impacto de cada componente introducido en la arquitectura del agente:  
- **El impacto de las nuevas heurísticas**: Tal y como sugiere la comparativa, añadir nuevas heurísticas mejoró el rendimiento de forma drástica. El modelo base de NeuralAgent (NeuralAgentDummy en nuestro código) presenta un desempeño deficiente, siendo incapaz de ganar una sola partida (tasa de victoria 0/10) y promediando puntuaciones fuertemente negativas (-386.7 en el mapa clásico), lo que indica muertes prematuras o bucles ineficientes. Al sumar las heurísticas personalizadas (NeuralAgent en nuestro código), la puntuación media se dispara a 636.8 puntos y se consiguen las primeras victorias (2/10). Esto demuestra que el modelo neuronal por sí solo carecía de conocimientos defensivos a corto plazo, una deficiencia que las heurísticas diseñadas logran mitigar. 

- **El impacto de la planificación (Alpha-Beta)**: La integración del algoritmo Minimax con Poda Alfa-Beta supuso el mayor salto cualitativo en el comportamiento del agente. Al abandonar la toma de decisiones pura del modo greedy para adoptar un razonamiento adversarial a múltiples pasos, la tasa de victorias se triplica en el entorno clásico (del 20% al 60%) y la puntuación promedio prácticamente se duplica, alcanzando los 1294.4 puntos. Esto confirma el objetivo principal de la práctica: la red neuronal ("intuición") se beneficia inmensamente de la capacidad de explorar un árbol de juego ("cálculo profundo").  

- **Capacidad de generalización al nuevo mapa**: Es fundamental recordar que la red neuronal fue entrenada exclusivamente en la configuración mediumClassic. A pesar de que el nuevo mapa (customMaze) altera las dinámicas del juego al contar con áreas más abiertas donde antes había paredes , el agente híbrido generaliza extraordinariamente bien. En el agente final (AlphaBeta + NN + heuristics), la penalización por el cambio de entorno es mínima: la tasa de éxito solo desciende ligeramente (de 6/10 a 5/10 victorias) y la puntuación se mantiene en niveles competitivos (1172.2). Esto se debe a que la Poda Alfa-Beta y las reglas heurísticas son agnósticas al diseño específico del mapa, dotando al sistema de la flexibilidad necesaria para desenvolverse en entornos desconocidos para la red. 

Debemos de tener en cuenta que los resultados obtenidos en la tabla previamente mostrada, han sido conseguidos con un valor de profundidad igual a 3. Con el objetivo de comprobar la mejora que obtendríamos en el nuevo mapa si subiésemos el nivel de profundidad, hemos vuelto a ejecutar la configuración de AlphaBeta + NN + heuristics con depth = 4.

In [13]:
new_layout_AlphaBeta_depth4= [1930.0, 1437.0, 2088.0, 918.0, 1745.0, 1903.0, 886.0, 1864.0, 851.0, 2114.0]

table2 = Texttable()  
table2.add_rows([["Configuration","New layout"],
               ["AlphaBeta + NN + heuristics (depth=4)",f"score: {statistics.mean(new_layout_AlphaBeta_depth4)} \nwin_rate: 7/10"]])

    
print(table2.draw())

+---------------------------------------+----------------+
|             Configuration             |   New layout   |
+=======================================+================+
| AlphaBeta + NN + heuristics (depth=4) | score: 1573.6  |
|                                       | win_rate: 7/10 |
+---------------------------------------+----------------+


- **Incremento en la tasa de victorias**: El porcentaje de éxito ascendió del 50% (5/10) al 70% (7/10). Este aumento del 20% demuestra que anticipar un turno completo más (el movimiento de Pacman y la respuesta correspondiente de los fantasmas) permite al agente esquivar situaciones de encierro que a profundidad 3 resultaban invisibles o inevitables.

- **Mejora de la puntuación media**: La puntuación media en el nuevo mapa se disparó de 1172.2 a 1573.6 puntos (un incremento de más de 400 puntos). Esto se traduce en que Pacman no solo sobrevive más, sino que es capaz de limpiar el mapa de comida de manera mucho más eficiente y rápida, maximizando las bonificaciones de tiempo y maximizando el consumo de fantasmas asustados cuando se presenta la oportunidad.

Aunque aumentar la profundidad incrementa el número de nodos en el árbol de juego de forma exponencial, la Poda Alfa-Beta implementada demostró ser lo suficientemente eficiente como para absorber el coste computacional extra sin ralentizar drásticamente el tiempo de respuesta por turno del agente.

## Parte 7. Comandos

Partida 1: AlphaBeta + NN + heuristics (depth=3) mapa original -> score: 2053

python pacman.py -p AlphaBetaNeuralAgent

Partida 2: AlphaBeta + NN + heuristics (depth=4) customLayout -> score: 1930

python pacman.py -l customMaze -p AlphaBetaNeuralAgent